api/clients/buckets/bucket.go

```diff
@@ -100,469 +39,63 @@ func WithRetrySettings(maxRetries int, durationBetweenTries time.Duration, maxWa
-	if err != nil {
-		return api.Response{}, fmt.Errorf("unable to delete object with bucket name %q: %w", bucketName, err)
-	}
-	if !resp.IsSuccess() {
-		return api.Response{StatusCode: resp.StatusCode, Data: resp.Payload, Request: resp.RequestInfo}, nil
-	}
-
-	// await bucket being successfully deleted
-	timoutCtx, cancel := context.WithTimeout(ctx, c.retrySettings.maxWaitDuration)
-	defer cancel() // cancel deadline if awaitBucketState returns before deadline
-	_, err = c.awaitBucketState(timoutCtx, bucketName, removed)
-	if err != nil {
-		return api.Response{}, fmt.Errorf("unable to delete object with bucket name %q: %w", bucketName, err)
-	}
-
-	return api.Response{StatusCode: resp.StatusCode, Data: resp.Payload, Request: resp.RequestInfo}, nil
-}
-
-// upsert is an internal function used by Upsert to perform the create or update logic.
-func (c Client) upsert(ctx context.Context, bucketName string, data []byte) (Response, error) {
-
-	logger := logr.FromContextOrDiscard(ctx)
-	ctx, cancel := context.WithTimeout(ctx, c.retrySettings.maxWaitDuration)
-	defer cancel()
-
-	// First, try to create a new bucket definition
-	resp, err := c.create(ctx, bucketName, data)
-	if err != nil {
-		return api.Response{}, err
-	}
-
-	// If creating the bucket definition worked, return the result
-	if resp.IsSuccess() {
-		logger.Info(fmt.Sprintf("Created bucket with bucket name %q", bucketName))
-		return api.Response{StatusCode: resp.StatusCode, Data: resp.Payload, Request: resp.RequestInfo}, nil
-	}
-
-	// Return if creation failed, but the errors was not 409 Conflict - Bucket already exists
-	if resp.StatusCode != http.StatusConflict {
-		return api.Response{StatusCode: resp.StatusCode, Data: resp.Payload, Request: resp.RequestInfo}, err
-	}
-
-	// If bucket is currently being deleted, retry upserts to re-create it once it's gone
-	if b, err := unmarshalJSON(&resp); err != nil && b.Status == "deleting" {
-		logger.V(1).Info(fmt.Sprintf("Bucket %q is being deleted. Waiting before re-creation...", b.BucketName))
-		time.Sleep(c.retrySettings.durationBetweenTries)
-		return c.upsert(ctx, bucketName, data)
-	}
-
-	// Try to update an existing bucket definition
-	logger.V(1).Info(fmt.Sprintf("Failed to create new object with bucket name %q. Trying to update existing object. API Error (HTTP %d): %s", bucketName, resp.StatusCode, resp.Payload))
-	apiResp, err := c.Update(ctx, bucketName, data)
-
-	if errors.Is(err, DeletingBucketErr) {
-		logger.V(1).Info(fmt.Sprintf("Failed to upsert bucket with name %q as it was being deleted. Re-creating...", bucketName))
-		if _, err := c.awaitBucketState(ctx, bucketName, removed); err != nil {
-			return Response{}, err
-		}
-		return c.Create(ctx, bucketName, data)
-	}
-	return apiResp, err
-}
-
-func (c Client) create(ctx context.Context, bucketName string, data []byte) (rest.Response, error) {
+func (c Client) Create(ctx context.Context, bucketName string, data []byte) (rest.Response, error) {
```

Comment:
The biggest question I have about this level of client is, what types should they return?

For now, I've stuck to `rest.Response`, as `api.Response` already needs some logic to be constructed.
This is not noticeable here, but will be for automation -> IMO a plain API client for automation should not resolve pagination and thus just return one rest.Response which might contain info about additional pages.

If these clients shouldn't use `api.Response` though, it should probably move up to the new `client/` package level too.